File: UnitTest/Benchmark/BBasicSimpleAddConcurrentWithConflict.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Zeze.Component;

namespace Benchmark
{
    [TestClass]
    public class BBasicSimpleAddConcurrentWithConflict
    {
        public const int AddCount = 100_0000;
        public const int ConcurrentLevel = 1_000;
        [TestMethod]
        public void testBenchmark() {
            demo.App.Instance.Start();
            try {
                demo.App.Instance.Zeze.NewProcedure(Remove, "remove").CallSynchronously();
                var tasks = new List<Task>(AddCount);
                for (int i = 1; i < ConcurrentLevel; ++i)
                {
                    int c = i;
                    tasks.Add(demo.App.Instance.Zeze.NewProcedure(async () =>
                    {
                        await demo.App.Instance.demo_Module1.Table1.GetOrAddAsync(c);
                        return 0;
                    }, "Add").CallAsync());
                }
                foreach (var task in tasks)
                    task.Wait();
                tasks.Clear();
                Console.WriteLine("benchmark start...");
                var b = new Zeze.Util.Benchmark();
                for (int i = 0; i < AddCount; ++i) {
                    var c = i % 1000;
                    tasks.Add(demo.App.Instance.Zeze.NewProcedure(async () => await Add(c), "Add").CallAsync());
                }
                b.Report(this.GetType().FullName, AddCount);
                foreach (var task in tasks) {
                    task.Wait();
                }
                b.Report(this.GetType().FullName, AddCount);
                demo.App.Instance.Zeze.NewProcedure(Check, "check").CallSynchronously();
                demo.App.Instance.Zeze.NewProcedure(Remove, "remove").CallSynchronously();
            }
            finally
            {
                demo.App.Instance.Stop();
            }
        }

        private async Task<long> Check() {
            var r = await demo.App.Instance.demo_Module1.Table1.GetOrAddAsync(0L);
            Assert.AreEqual(r.Long2, AddCount);
            //System.out.println(r.getLong2());
            return 0;
        }

        private async Task<long> Add(long key) {
            var r = await demo.App.Instance.demo_Module1.Table1.GetOrAddAsync(0L);
            r.Long2 += 1;
            //for (int i = 0; i < 100000; ++i)
            //    r.Int_1 += i;
            return 0;
        }

        private async Task<long> Remove() {
            await demo.App.Instance.demo_Module1.Table1.RemoveAsync(0L);
            return 0;
        }
    }
}

File: ZezeJava/ZezeJavaTest/src/Benchmark/BBasicSimpleAddConcurrentWithConflict.java
package Benchmark;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.concurrent.Future;
import Zeze.Util.PerfCounter;
import Zeze.Util.TaskSpec;
import demo.App;
import org.junit.jupiter.api.Assertions;

@SuppressWarnings("NewClassNamingConvention")
public class BBasicSimpleAddConcurrentWithConflict {
	public static final int AddCount = 10_000;

	@Test

	public void testBenchmark() throws Exception {
		App.Instance.Start();
		try {
			App.Instance.Zeze.newProcedure(BBasicSimpleAddConcurrentWithConflict::Remove, "remove").call();
			ArrayList<Future<Long>> tasks = new ArrayList<>(AddCount);
			System.out.println("benchmark start...");
			var b = new Zeze.Util.Benchmark();
			for (int i = 0; i < AddCount; ++i) {
				tasks.add(TaskSpec.ofProcedure(App.Instance.Zeze.newProcedure(BBasicSimpleAddConcurrentWithConflict::Add, "Add")).submitNow());
				if ((i + 1) % 10 == 0) {
					for (var task : tasks)
						task.get();
					tasks.clear();
				}
			}
			//b.Report(this.getClass().getName(), AddCount);
			for (var task : tasks) {
				task.get();
			}
			b.report(this.getClass().getName(), AddCount);
			System.out.println(PerfCounter.instance().getLogAndReset());
			App.Instance.Zeze.newProcedure(BBasicSimpleAddConcurrentWithConflict::Check, "check").call();
			App.Instance.Zeze.newProcedure(BBasicSimpleAddConcurrentWithConflict::Remove, "remove").call();
		} finally {
			//App.Instance.Stop();
		}
	}

	private static long Check() {
		var r = App.Instance.demo_Module1.getTable1().getOrAdd(1L);
		Assertions.assertEquals(AddCount, r.getLong2());
		//System.out.println(r.getLong2());
		return 0;
	}

	private static long Add() {
		var r = App.Instance.demo_Module1.getTable1().getOrAdd(1L);
		r.setLong2(r.getLong2() + 1);
		return 0;
	}

	private static long Remove() {
		App.Instance.demo_Module1.getTable1().remove(1L);
		return 0;
	}
}
